services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/tormaystarkastelu/TormaystarkasteluTormaysService.kt

```diff
@@ -89,64 +87,40 @@ class TormaystarkasteluTormaysService(private val jdbcOperations: JdbcOperations
     fun getIntersectingBusRoutes(geometriaIds: Set<Int>): Set<TormaystarkasteluBussireitti> {
         if (geometriaIds.isEmpty()) return setOf()
         val placeholders = Collections.nCopies(geometriaIds.size, "?").joinToString(", ")
-        with(jdbcOperations) {
-            // TODO: DISTINCT ON (route_id, direction_id) at the database level
-            return query(
-                    """
-                SELECT 
-                    tormays_buses_polys.fid,
-                    tormays_buses_polys.route_id,
-                    tormays_buses_polys.direction_id,
-                    tormays_buses_polys.rush_hour,
-                    tormays_buses_polys.trunk,
-                    hankegeometria.id
-                FROM tormays_buses_polys, hankegeometria
-                WHERE hankegeometria.hankegeometriatid in ($placeholders)
-                AND ST_Intersects(tormays_buses_polys.geom, hankegeometria.geometria)
-                """
-                        .trimIndent(),
-                    { rs, _ ->
-                        TormaystarkasteluBussireitti(
-                            rs.getString(2),
-                            rs.getInt(3),
-                            rs.getInt(4),
-                            TormaystarkasteluBussiRunkolinja.valueOfRunkolinja(rs.getString(5))
-                        )
-                    },
-                    *geometriaIds.toTypedArray()
-                )
-                .toHashSet()
-        }
+        val sql =
+            """
+            SELECT DISTINCT ON (buses.route_id, buses.direction_id)
+                buses.route_id,
+                buses.direction_id,
+                buses.rush_hour,
+                buses.trunk
+            FROM tormays_buses_polys buses, hankegeometria
+            WHERE hankegeometria.hankegeometriatid in ($placeholders)
+            AND ST_Intersects(buses.geom, hankegeometria.geometria)
+            ORDER BY buses.route_id, buses.direction_id, buses.trunk DESC, buses.rush_hour DESC
```

Comment:
Do we need ordering?

Reply:
Well, it does not make so much difference in performance that it would matter - I was mostly curious why use `ORDER BY` here when the order does not matter as the results are returned in an unordered `Set`.

Reply:
`DISTINCT ON (buses.route_id, buses.direction_id)` needs the rows to be ordered by `route_id` and `direction_id`, and it selects the first row with unique values of the parameters. So we sort them by `trunk` and `rush_hour`, selecting the worst case for each route and direction.

Reply:
I see. If I run these two sql clauses in our dev environment they both return the exactly same results except for one row:
```
SELECT DISTINCT ON (buses.route_id, buses.direction_id)
    buses.route_id,
    buses.direction_id,
    buses.rush_hour,
    buses.trunk
FROM tormays_buses_polys buses, hankegeometria
WHERE hankegeometria.hankegeometriatid in (select distinct hankegeometriatid from hankegeometria)
  AND ST_Intersects(buses.geom, hankegeometria.geometria) 
ORDER BY buses.route_id, buses.direction_id, buses.trunk DESC, buses.rush_hour DESC;
```

```
SELECT DISTINCT ON (buses.route_id, buses.direction_id)
    buses.route_id,
    buses.direction_id,
    buses.rush_hour,
    buses.trunk
FROM tormays_buses_polys buses, hankegeometria
WHERE hankegeometria.hankegeometriatid in (select distinct hankegeometriatid from hankegeometria)
  AND ST_Intersects(buses.geom, hankegeometria.geometria);
```

The difference is this:
```
19c19
<  1024     |            0 |         3 | no
---
>  1024     |            0 |         0 | no
```

Without `ORDER BY` it does not necessarily get the max `rush_hour`. 

Goog to know this, thanks!